cli/toolchain_cmd.go

```diff
@@ -522,25 +616,75 @@ func installPythonToolchain() error {
 }
 
 func installJavaToolchain() error {
-	const toolchain = "sourcegraph.com/sourcegraph/srclib-java"
+	return installToolchainFromBundle("Java", "sourcegraph.com/sourcegraph/srclib-java", "https://srclib-support.s3-us-west-2.amazonaws.com/srclib-java/srclib-java__bundle__jdk-1.8.tar.gz")
+}
 
-	requiredCmds := map[string]string{
-		"javac": "install the JDK (JDK8u preferred)",
+func installToolchainFromBundle(name, toolchainPath, bundleURL string) (err error) {
+	tmpDir, err := ioutil.TempDir("", "srclib-toolchain-bundle")
+	if err != nil {
+		return err
 	}
-	for requiredCmd, instructions := range requiredCmds {
-		if _, err := exec.LookPath(requiredCmd); isExecErrNotFound(err) {
-			return skippedToolchain{toolchain, fmt.Sprintf("no `%s` found in PATH; to install, %s", requiredCmd, instructions)}
+
+	url, err := url.Parse(bundleURL)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("Installing %s toolchain from %s", name, bundleURL)
+	outputFile := filepath.Join(tmpDir, filepath.Base(url.Path))
+	f, err := os.Create(outputFile)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err2 := f.Close(); err2 != nil && err == nil {
+			err = err2
 		}
+	}()
+
+	resp, err := http.Get(bundleURL)
+	if err != nil {
+		return err
 	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("GET %s: unexpected HTTP status %d (%s)", bundleURL, resp.StatusCode, resp.Status)
+	}
+
+	done := make(chan struct{})
+	go func() {
+		for {
+			t := time.After(time.Second * 2)
+			select {
+			case <-t:
+				fi, err := os.Stat(outputFile)
+				if err == nil {
+					fmt.Printf("\rDownload %.1f%% complete (%.1f MB / %.1f MB)",
+						float64(fi.Size())/float64(resp.ContentLength)*100,
+						float64(fi.Size())/1024/1024,
+						float64(resp.ContentLength)/1024/1024,
+					)
+				}
+			case <-done:
+				log.Printf("Finished downloading")
+				return
+			}
+		}
+	}()
 
-	srclibpathDir := filepath.Join(strings.Split(srclib.Path, ":")[0], toolchain) // toolchain dir under SRCLIBPATH
-	log.Println("Downloading or updating Java toolchain in", srclibpathDir)
-	if err := execSrcCmd("toolchain", "get", "-u", toolchain); err != nil {
+	log.Printf("Downloading %s (%.1f MB)", bundleURL, float64(resp.ContentLength)/1024/1024)
+	_, err = io.Copy(f, resp.Body)
+	done <- struct{}{}
```

Comment:
Pretty minor, I would move the `log.Printf("Finished downloading")` line from the goroutine above to here. Otherwise, it may end up printing a little later (since this is a one-way synchronization) and looking odd:

```
Downloading https://srclib-support.s3-us-west-2.amazonaws.com/...tar.gz (356.4 MB)
Download 99.6% complete (354.9 MB / 356.4 MB)
Unarchiving Java toolchain bundle at /var/folders/_w/3kl_g4rn5sv1nyjblk7tf4h00000gn/T/srclib-toolchain-bundle402629058/srclib-java__bundle__jdk-1.8.tar.gz
Finished downloading
```